wasmtime/_table.py

```diff
@@ -53,65 +44,75 @@ def type(self) -> TableType:
 
         Raises a `WasmtimeError` if the table could not be grown.
         Returns the previous size of the table otherwise.
         """
-        init_ptr = get_func_ptr(init)
-
-        prev = c_uint32(0)
-        error = ffi.wasmtime_funcref_table_grow(self._ptr, c_uint32(amt), init_ptr, byref(prev))
-        if error:
+        init_val = Val._convert(self.type.element, init)
+        ok = ffi.wasm_table_grow(self._ptr, c_uint32(amt), init_val._raw.of.ref)
+        if not ok:
             raise WasmtimeError("failed to grow table")
-        return prev.value
+        return self.size - amt
 
-    def __getitem__(self, idx: int) -> Optional[Func]:
+    def __getitem__(self, idx: int) -> Optional[Any]:
         """
-        Gets an individual element within this table. Currently only works on
-        `funcref` tables.
+        Gets an individual element within this table.
+
+        Returns `None` for null references in the table (i.e. a null `funcref`
+        or a null `externref).
+
+        Returns a `Func` for non-null `funcref` table elements.
+
+        Returns the wrapped extern data for non-null `externref` table elements.
 
-        Returns `None` for a slot where no function has been placed into.
-        Returns `Func` for a slot with a function.
         Raises an `WasmtimeError` if `idx` is out of bounds.
         """
+        if idx >= self.size:
+            raise WasmtimeError("table index out of bounds")
 
-        ptr = POINTER(ffi.wasm_func_t)()
-        ok = ffi.wasmtime_funcref_table_get(self._ptr, idx, byref(ptr))
-        if ok:
-            if ptr:
-                return Func.__from_ptr__(ptr, None)
-            return None
-        raise WasmtimeError("table index out of bounds")
+        if self.type.element == ValType.externref():
+            val = Val.externref(None)
+        elif self.type.element == ValType.funcref():
+            val = Val.funcref(None)
+        else:
+            raise WasmtimeError("unsupported table element type")
 
-    def __setitem__(self, idx: int, val: Optional[Func]) -> None:
+        val._raw.of.ref = ffi.wasm_table_get(self._ptr, idx)
+        return val.value
+
+    def __setitem__(self, idx: int, val: IntoVal) -> None:
         """
-        Sets an individual element within this table. Currently only works on
-        `funcref` tables.
+        Sets an individual element within this table.
+
+        `idx` must be an integer index.
 
-        The `val` specified must either be a `Func` or `None`, and `idx` must
-        be an integer index.
+        The `val` specified must be convertible into this table's element
+        type. I.e. for a `funcref` table, `val` must either be a `Func` or
+        `None`, and for an `externref` table, `val` may be any arbitrary
+        external data.
 
         Raises a `WasmtimeError` if `idx` is out of bounds.
         """
+        if idx >= self.size:
+            raise WasmtimeError("Index out of bounds when setting table element")
```

Comment:
This should be part of the `wasm_table_set` API I think?

Reply:
Or if this is for a better error message perhaps this could happen in the `not ok` branch?

Reply:
Yeah, the `wasm_table_get` signature doesn't differentiate between kinds of failures or provide any sort of message, so this is doing an explicit check to provide the better error message.